### ISSUE TITLE
recursor: add option to validate dnssec and log bogus but do not send servfail to client

### ISSUE THREAD

**Issue by CaptainQwark** [PowerDNS/pdns#4436]
It would be nice if it was possible to let the recursor validate DNSSEC and log bogus validations, but never actually send a SERVFAIL to any client (even those who have the AD bit set).

use case:
For an ISP it's crucial to know what the impact will be of turning on validation **before** actually turning it on and have you customers impacted by failing dns queries.
By just validating and logging without servfails we can see what domains that our customers request are bogus and:
- validate that the recursor is correct about the bogus validation
- address the issue with the domain admins of the failing domains to get things fixed before we start validating (I know there will always be crappy configured domains, but anything helps)
- determine if it's feasible at any point in time to turn on validation for our customers

**Comment by pieterlexis** [comment]
So I am pondering the following 3 options to replace `dnssec` and `dnssec-log-bogus`:

- `dnssec-validation`
    - `yes` -- Always validate
    - `client` -- Validate when the client requests it
    - `no` -- Disable DNSSEC validation
- `dnssec-servfail`
    - `yes` -- Always SERVFAIL on bogus
    - `client` -- Only when the client requests it
    - `no` -- Never send a SERVFAIL on bogus
- `dnssec-log`
  - `yes` or `all` -- Log all validation results
  - `bogus` -- Log all bogus results
  - `no` -- Don't log validation results

This will allow for greater operator flexibility without having too many options to be confusing.

**Comment by pieterlexis** [comment]
Alternatively, the following might also work, but might also be pretty confusing:

- `dnssec`
    - `yes` -- Always validate and SERVFAIL
    - `yes-client` -- Always validate and SERVFAIL if the client asks for DNSSEC
    - `yes-nofail` -- Always validate but don't SERVFAIL
    - `client` -- Only validate when the client asks for it
    - `client-nofail` -- Validate when the client asks for it, but don't SERVFAIL
    - `no` -- Disable DNSSEC validation
- `dnssec-log`
  - `yes` or `all` -- Log all validation results
  - `bogus` -- Log all bogus results
  - `no` -- Don't log validation results

**Comment by CRtEurope** [comment]
The first proposal from pieterlexis is better and not confusing at all. The second one, hmmm, is confusing.

**Comment by CaptainQwark** [comment]
+1 for the first proposal. The only downside may be that it breaks current dnssec configurations.

**Comment by paddg** [comment]
Yes, first proposal sounds good! Maybe useless combinations will be possible:
_Always validate but Never send a SERVFAIL on bogus_

**Comment by pieterlexis** [comment]
>  Always validate but Never send a SERVFAIL on bogus

combined with a log, it makes sense.. but on it's own.. not really.. but hey :smile:

**Comment by CaptainQwark** [comment]
@paddg combined with logging, that combination is exactly why I created this request :)

**Comment by paddg** [comment]
@CaptainQwark 9 switches for 1 feature, must be DNSSEC 😀
But you're right of course!

**Comment by nottledim** [comment]
It seems to me, in my ignorance, that a setting where validation only takes place for sites which are dnssec configured is needed.  Since the vast majority of sites at the moment do not have dnssec configured most of the world is bogus.  Dnssec will never get off the ground if SERVFAIL is returned for unconfigured sites, especially so since resolvers have no way to provide a secure/insecure status for returned addresses.

**Comment by johnhtodd** [comment]
nottledim: SERVFAIL is not sent on unconfigured sites, so this isn't a concern.

On the original query: if you’re looking to get a list of “Bogus” zones but don’t want to fail on them, why not just set up another resolver and use the “TeeAction” (using dnsdist in front of your primary resolver) your queries to it? http://dnsdist.org/README/#teeaction  You could even point that second resolver back at your primary resolver so you’d be using (mostly) cache and not increasing your outbound queries (hint: build in a delay before #2 asks #1)
